### ISSUE TITLE
Missing sum in Epistemic and Aleatoric computation

### ISSUE THREAD

**Issue by Zimba96** [kumar-shridhar/PyTorch-BayesianCNN#67]
Hey there,

I recently tried to implement a Bayesian Net myself and oriented myself a little on your implementation. When coming across the computation of the epistemic and aleatoric, I noticed that your implementation in uncertainty_estimation.py seems to differ a little from the formula in your paper.
In the paper, it says:
![formula_paper](https://user-images.githubusercontent.com/38334285/121496635-f2e3c900-c9da-11eb-93f7-7085bb0e3e52.png)

However, I'm missing the sums in your implementation, e. g. when you calculate the epistemic and you divide by T but not summing over all t's (same with the second part of the aleatoric):
```
temp = p_hat - np.expand_dims(p_bar, 0)
epistemic = np.dot(temp.T, temp) / T
epistemic = np.diag(epistemic)

aleatoric = np.diag(p_bar) - (np.dot(p_hat.T, p_hat) / T)
aleatoric = np.diag(aleatoric)
```
Or does the following np.diag somehow replaces the summation?

There's another short question I'd like to ask:
This code outputs the epistemic and aleatoric as vectors of size [number_of_classes], right? Does that mean that the vector represents the epistemic and aleatoric for the predictions of each class? And when I want to obtain the epistemic and aleatoric of my resulting prediction (the class with the max. softmax/softplus-value), do I then only need to extract the aleatoric and epistemic-value corresponding to that class from the epistemic and aleatoric vectors?

Thanks in advance for your answers!

**Comment by Zimba96** [comment]
For the first part of my question, I figured it out myself. Instead of throwing the input image into the net T times separately and sample everytime, one input data consisting T times the original input data is given to the net:
```
input_image = input_image.unsqueeze(0)
input_images = input_image.repeat(T, 1, 1, 1)
net_out, _ = model(input_images)
```
The summations given in the Var-equation (15) is then contained in the dot-product, so that instead of having this equation (like it would be according to equation 15):
![grafik](https://user-images.githubusercontent.com/38334285/124552752-398ade80-de34-11eb-8aa5-33aacb59db14.png)
we get this one:
![grafik](https://user-images.githubusercontent.com/38334285/124553377-f1b88700-de34-11eb-8036-0c5320cfd02b.png)
which is basically the same.

For the second part of my question, I figured out that the epistemic uncertainty is a scalar vector while the aleatoric uncertainty is kind of a covariance matrix. What I haven't figured out yet is how to deal with this covariance matrix since I assumed that the aleatoric uncertainty also is just a scalar value representing the variance within the given data. Is there a way to obtain this single-valued aleatoric uncertainty out of this covariance matrix or did I get the idea behind the aleatoric uncertainty wrong?
